prusti-encoder/src/encoders/mir_impure.rs

```diff
@@ -439,59 +433,27 @@ impl<'tcx, 'vir, 'enc> EncoderVisitor<'tcx, 'vir, 'enc> {
+        for &elem in place.projection {
+            expr = self.encode_place_element(place_ty, elem, expr);
+            place_ty = place_ty.projection_ty(self.vcx.tcx, elem);
+        }
+        expr
     }
 
-    // TODO: this will not work for unevaluated constants (which needs const
-    // MIR evaluation, more like pure fn body encoding)
-    fn encode_constant(
-        &self,
-        constant: &mir::Constant<'tcx>,
-    ) -> vir::Expr<'vir> {
-        match constant.literal {
-            mir::ConstantKind::Val(const_val, const_ty) => {
-                match const_ty.kind() {
-                    ty::TyKind::Tuple(tys) if tys.len() == 0 =>
-                        self.vcx.mk_todo_expr(vir::vir_format!(self.vcx, "s_Tuple0_cons()")),
-                    ty::TyKind::Int(int_ty) => {
-                        let scalar_val = const_val.try_to_scalar_int().unwrap();
-                        self.vcx.mk_todo_expr(
-                            vir::vir_format!(
-                                self.vcx,
-                                "s_Int_{}_cons({})",
-                                int_ty.name_str(),
-                                scalar_val.try_to_int(scalar_val.size()).unwrap()
-                            )
-                        )
-                    }
-                    ty::TyKind::Uint(uint_ty) => {
-                        let scalar_val = const_val.try_to_scalar_int().unwrap();
-                        self.vcx.mk_todo_expr(
-                            vir::vir_format!(
-                                self.vcx,
-                                "s_Uint_{}_cons({})",
-                                uint_ty.name_str(),
-                                scalar_val.try_to_uint(scalar_val.size()).unwrap()
-                            )
-                        )
-                    }
-                    ty::TyKind::Bool => self.vcx.mk_todo_expr(
-                        vir::vir_format!(
-                            self.vcx,
-                            "s_Bool_cons({})",
-                            const_val.try_to_bool().unwrap()
-                        )
-                    ),
-                    unsupported_ty => todo!("unsupported constant literal type: {unsupported_ty:?}"),
-                }
+    fn encode_place_element(&mut self, place_ty: mir::tcx::PlaceTy<'tcx>, elem: mir::PlaceElem<'tcx>, expr: vir::Expr<'vir>) -> vir::Expr<'vir> {
+         match elem {
+            mir::ProjectionElem::Field(field_idx, _) => {
+                let e_ty = self.deps.require_ref::<PredicateEnc>(place_ty.ty).unwrap();
+                let field_access = e_ty.expect_variant_opt(place_ty.variant_index).ref_to_field_refs;
+                let projection_p = field_access[field_idx.as_usize()];
+                projection_p.apply(self.vcx, [expr])
             }
-            unsupported_literal => todo!("unsupported constant literal: {unsupported_literal:?}"),
+            // All variants start at the same `Ref`?
```

Comment:
Is this a TODO or comment about what's happening?

Reply:
TODO, updated